### ISSUE TITLE
This application is modifying the autolayout engine from a background thread after the engine was accessed from the main thread. This can lead to engine corruption and weird crashes.

### ISSUE THREAD

**Issue by StephanKornerTrihow** [firebase/firebase-ios-sdk#6746]
### [REQUIRED] Step 1: Describe your environment

  * Xcode version: 12.0.1
  * Firebase SDK version: 6.20 works well / version: 6.34 crashes
  * Firebase Component: 
  pod 'Firebase/Auth', '6.20.0'  # v6.34 crashes
  pod 'Firebase/Core', '6.20.0' # v6.34 crashes
  pod 'Firebase/Database', '6.20.0' # v6.34 crashes
  pod 'Firebase/Firestore', '6.20.0' # v6.34 crashes
  pod 'Firebase/Functions', '6.20.0' # v6.34 crashes
  pod 'Firebase/Messaging', '6.20.0' # v6.34 crashes
  pod 'Firebase/Performance', '6.20.0' # v6.34 crashes
  pod 'Firebase/RemoteConfig', '6.20.0' # v6.34 crashes
  pod 'Firebase/Storage', '6.20.0' # v6.34 crashes
 
  pod 'FirebaseUI/Firestore', '8.5.1'
  pod 'FirebaseUI/Storage', '8.5.1'

  * Installation method: CocoaPods

### [REQUIRED] Step 2: Describe the problem

My iOS App works well under iOS14.0.1 and Firebase-Components v6.20.

However, as soon as I update to Firebase-Components v6.34, then I get App crashes.

The error message at such a crash sais:
`This application is modifying the autolayout engine from a background thread after the engine was accessed from the main thread. This can lead to engine corruption and weird crashes.`

I get the very same error message when I prepare a situation where I do UI-work on a background thread.
Or in other words, if I forget to
A) `dispatchqueue.main.async { ... }` to the main thread
B) have completion blocks that work with `self` and I forget to weak-capture self in the first place, such as
```Swift
dispatchqueue.main.async { [weak self] in 
    guard let self = self else { return }
}
```

#### Steps to reproduce:

Using iOS14.0.1, XCode 12.0.1, Swift 5.3
--> Start my App and use Firebase v6.34 crashes the App.
--> Start my App and use Firebase v6.20 does work well.

I suspect that Google did forget to consider Apple's iOS14 restriction on background-thread usage of the UI.
As I read, Apple did restrict this policy further. And the Google Firebase iOS-SDK did probably not yet catch up with those Apple-changes.

Can you please tell me if you recognized the issue and by when we can expect an improved Firebase-SDK for iOS14 ?

**Comment by morganchen12** [comment]
None of the SDKs you've listed have any autolayout code in them. Check your callback handlers are being invoked on the main thread. We'll need a reproducible example to continue investigating this issue.

**Comment by StephanKornerTrihow** [comment]
Thank you @morganchen12 , for your quick reply. 

In general, I am just wondering why Firebase v6.20 in our App does not bring any error, however v6.34 does. So it must have something to do with the SDK (not necessarily with our callbacks). Or how do you explain the difference between behaviour for the two SDK's. (all under iOS14.0.1, Swift5.3).

To reply your example: 

I tried to find the place where the error happens but following any of these Swift- or XCode-techniques ([shown here](https://stackoverflow.com/questions/28302019/getting-a-this-application-is-modifying-the-autolayout-engine-from-a-background)) did not help. I don't know how to place an exception breakpoint for Firebase SDK added by Cocoapods. Therefore it is very hard to tell which callback is causing the error.

I can tell that the error happens sometimes already at startup. 

Here is what our App does at startup:

```Swift
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {

        // Cloud logging
        if Config.Firebase.Remote.bool(forKey: .swiftyBeaverCloud, defaultValue: true) {
            let cloudDestination = SBPlatformDestination(appID: Config.SwiftyBeaver.appID, appSecret: Config.SwiftyBeaver.appSecret, encryptionKey: Config.SwiftyBeaver.encryptionKey)
            log.addDestination(cloudDestination)
            self.cloudDestination = cloudDestination
        }
                
        // Configure cache
        SDImageCache.shared.config.maxDiskAge = Config.Firebase.Remote.double(forKey: .imageCacheMaxAge, defaultValue: 2592000)
        
        FirebaseApp.configure()
        
        // Use Firebase Functions Emulator in DEBUG mode if set
        #if DEBUG && false
        log.info("Enable Firebase Functions Emulator: \(origin)")
        Functions.shared.useFunctionsEmulator(origin: origin)
        #endif
        
        // Sign-out if running UI tests
        if Config.isRunningUITests {
            try? Auth.auth().signOut()
        }
        
        // Get notifications about auth state
        Auth.auth().addStateDidChangeListener { [weak self] auth, user in
            self?.authStateDidChange(user: user)
        }
        
        // Configure Remote Config
        RemoteConfig.remoteConfig().setDefaults(fromPlist: "RemoteConfigDefaults")
        self.fetchRemoteConfig()
        
        // Autologin existing users
        if let _ = Auth.auth().currentUser {
            self.changeRootViewController(.sessions, animated: false)
        }
         
        return true
    }
```

Other than that, we use Firebase listeners in our ViewControllers. An example is here :

```Swift
import UIKit
import FirebaseAuth
import FirebaseFirestore
import FirebaseFunctions
import FirebaseUI

class MyViewController: UIVieweController {
    
    var listener: ListenerRegistration?
    private var authListenerHandle: AuthStateDidChangeListenerHandle?
    
    deinit {
        listener?.remove()
        if let handle = authListenerHandle {
            Auth.auth().removeStateDidChangeListener(handle)
        }
    }

    override func viewDidLoad() {

        if let uid = Auth.auth().currentUser?.uid {
            self.myUID = uid
            
            // Listen for authentication changes
            self.authListenerHandle = Auth.auth().addStateDidChangeListener { [weak self] (auth, user) in
                if user?.uid != uid {
                    self?.listener?.remove()
                    self?.listener = nil
                    log.info("Stopped realtime updates for sessions")
                    
                    if let handle = self?.authListenerHandle {
                        Auth.auth().removeStateDidChangeListener(handle)
                    }
                }
            }
            
            // Build the session query
            let query = Firestore.firestore().collection(Config.Firebase.Collection.mycollection)
                .whereField(Config.Firebase.Field.myfield, arrayContains: uid)
                .order(by: Config.Firebase.Field.startTime, descending: true)
                .limit(to: 1000)
            
            // Listen for changes
            self.listener = query.addSnapshotListener { [weak self] (snapshot, error) in
                
                // Process new data
                let newData: [MyCollection] = snapshot?.values() ?? []
                
                // ...
            }
        }
     }
  }
```

**Comment by morganchen12** [comment]
If you can set a breakpoint for UIKit accesses on a background thread and share the stack trace, that would be helpful. It's not clear what's accessing AutoLayout from the code you've shared.

**Comment by StephanKornerTrihow** [comment]
Here is the stack:

```Swift
22:41:19.303 💙 INFO AppDelegate.fetchRemoteConfig():558 - Remote config fetched successfully
22:41:19.304 💙 INFO AppDelegate.fetchRemoteConfig():569 - Remote config activated
22:41:19.416 💙 INFO AppDelegate.messaging():318 - FCM registration token: abcde-se3reasdfasdfCCDFSAFS-dsafj ADASDDAFSD-?DFsdasdar_adfasfas_dsae123dasd?_dfasfasobfuscated
2020-10-19 22:41:20.952469+0200 Mycompany[26959:4840568] [connection] nw_resolver_start_query_timer_block_invoke [C6] Query fired: did not receive all answers in time for app-measurement.com:443
2020-10-19 22:41:21.041604+0200 Mycompany[26959:4840628] [connection] nw_resolver_start_query_timer_block_invoke [C8] Query fired: did not receive all answers in time for firebaseremoteconfig.googleapis.com:443
2020-10-19 22:41:21.042154+0200 Mycompany[26959:4840628] [connection] nw_resolver_start_query_timer_block_invoke [C9] Query fired: did not receive all answers in time for firebaseremoteconfig.googleapis.com:443
2020-10-19 22:41:21.052490+0200 Mycompany[26959:4840628] [connection] nw_resolver_start_query_timer_block_invoke [C7] Query fired: did not receive all answers in time for fcmtoken.googleapis.com:443
2020-10-19 22:41:21.140698+0200 Mycompany[26959:4840628] [connection] nw_resolver_start_query_timer_block_invoke [C10] Query fired: did not receive all answers in time for app-measurement.com:443
2020-10-19 22:41:21.320320+0200 Mycompany[26959:4840628] [connection] nw_resolver_start_query_timer_block_invoke [C11] Query fired: did not receive all answers in time for api.swiftybeaver.com:443
2020-10-19 22:41:21.435462+0200 Mycompany[26959:4840568] [connection] nw_resolver_start_query_timer_block_invoke [C12] Query fired: did not receive all answers in time for iid.googleapis.com:443
2020-10-19 22:41:24.966436+0200 Mycompany[26959:4840569] This application is modifying the autolayout engine from a background thread after the engine was accessed from the main thread. This can lead to engine corruption and weird crashes.
 Stack:(
	0   CoreAutoLayout                      0x00000001a1feb56c 3BC7CF7C-2E75-3882-9037-E607C116B9CA + 62828
	1   CoreAutoLayout                      0x00000001a1feb928 3BC7CF7C-2E75-3882-9037-E607C116B9CA + 63784
	2   UIKitCore                           0x00000001911aae78 00EA1426-38F7-3FD2-BE01-04EBD44ECA35 + 15957624
	3   UIKitCore                           0x00000001912998a0 00EA1426-38F7-3FD2-BE01-04EBD44ECA35 + 16935072
	4   UIKitCore                           0x00000001903a87e0 00EA1426-38F7-3FD2-BE01-04EBD44ECA35 + 1267680
	5   QuartzCore                          0x00000001917999e8 3E647DA5-5843-3774-B65A-A1928066EBBC + 1395176
	6   QuartzCore                          0x000000019179feb4 3E647DA5-5843-3774-B65A-A1928066EBBC + 1420980
	7   QuartzCore                          0x00000001917ab1d0 3E647DA5-5843-3774-B65A-A1928066EBBC + 1466832
	8   QuartzCore                          0x00000001916f7500 3E647DA5-5843-3774-B65A-A1928066EBBC + 730368
	9   QuartzCore                          0x0000000191721698 3E647DA5-5843-3774-B65A-A1928066EBBC + 902808
	10  QuartzCore                          0x0000000191721ea4 3E647DA5-5843-3774-B65A-A1928066EBBC + 904868
	11  libsystem_pthread.dylib             0x00000001d3dab110 9A318917-27DB-312E-91D6-81661034FCC6 + 57616
	12  libsystem_pthread.dylib             0x00000001d3da6bcc 9A318917-27DB-312E-91D6-81661034FCC6 + 39884
	13  libsystem_pthread.dylib             0x00000001d3da88a0 9A318917-27DB-312E-91D6-81661034FCC6 + 47264
	14  libsystem_pthread.dylib             0x00000001d3da8634 _pthread_wqthread + 416
	15  libsystem_pthread.dylib             0x00000001d3dab874 start_wqthread + 8
)
Modifications to the layout engine must not be performed from a background thread after it has been accessed from the main thread.
```

No idea which Firebase module causes it... But it all hapens 3 seconds after the App sucessfully startsup.

**Comment by paulb777** [comment]
We've have some reports of the most recent FirebasePerformance release causing crashes like #6734. Does removing Firebase/Performance or going back to 6.33.0 make a difference?

**Comment by morganchen12** [comment]
The stack trace doesn't show any symbols from Firebase or your app module (and the non-app module symbols are not deobfuscated). Can you try to identify what's enqueueing this background thread work, or share your project?

**Comment by StephanKornerTrihow** [comment]
Thank you @paulb777 , I will check later....since I just found the causing spot in my code. Maybe you have an explanation why SDK 6.34 makes a difference towards 6.20:

In my AppDelegate's `didFinishLaunchingWithOptions` mehtod I call the following method :

```Swift
private func fetchRemoteConfig() {
        var expiration: TimeInterval = 43200.0 // 12 hours
        if UserDefaults.standard.bool(forKey: Config.Defaults.remoteConfigStaleKey) {
            expiration = 0;
        }
        RemoteConfig.remoteConfig().fetch(withExpirationDuration: expiration) { (status, error) in
            switch status {
            case .noFetchYet:
                log.info("Remote config has not been fetched yet")
            case .throttled:
                log.info("Remote config fetch is throttled")
            case .failure:
                log.error("Remote config fetch failed: \(error?.localizedDescription ?? "-")")
            case .success:
                log.info("Remote config fetched successfully")
                UserDefaults.standard.set(false, forKey: Config.Defaults.remoteConfigStaleKey)
                RemoteConfig.remoteConfig().activate { (error) in
                    if let error = (error as NSError?) {
                        // Don't log the following error:
                        // Remote config activation failed: Error Domain=com.google.remoteconfig.ErrorDomain Code=8003 "(null)" UserInfo={ActivationFailureReason=Most recently fetched config is already activated}
                        if error.code != 8003 {
                            log.error("Remote config activation failed: \(error)")
                        }
                    }
                    else {
                        log.info("Remote config activated")
                        NotificationCenter.default.post(name: Config.Firebase.Remote.updateNotification, object: self)
                    }
                }
            @unknown default:
                break
            }
        }
    }
```

From what I see, in there, the `NotificationCenter.default.post(name: Config.Firebase.Remote.updateNotification, object: self)` is called.

And this notifcation has the following observer trigger a UICollectionViewFlowLayout update using `self.invalidateLayout()`

```Swift
class MediaFlowLayout: UICollectionViewFlowLayout {
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    override init() {
        super.init()
        NotificationCenter.default.addObserver(self, selector: #selector(self.remoteConfigDidUpdate(_:)), name: Config.Firebase.Remote.updateNotification, object: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        NotificationCenter.default.addObserver(self, selector: #selector(self.remoteConfigDidUpdate(_:)), name: Config.Firebase.Remote.updateNotification, object: nil)
    }
    
    @objc func remoteConfigDidUpdate(_ notif: Notification) {
        self.invalidateLayout()
    }

   //...
}
```

AND THIS `invalidateLayout()` is causing the App crash since it is most likely not on a main-thread.

I am very curious to find out why Firebase SDK 6.20 does not crash but  SDK 6.34 does, given the code above !!!
Any idea ?

Oh, and I changed two more things in my code :

A) I tried to follow the deprication-warning in the `RemoteConfig.remoteConfig.activate` call and changed the success case from `RemoteConfig.remoteConfig().activate { (error) in ....` to `RemoteConfig.remoteConfig().activate { (bool, error) in ...`

But that did not change much.

B) And secondly I added `DispatchQueue.main.async { ... }` arround the 'self.invalidateLayout()` in the NotificationCenter-observer.

--> AND THIS SEEMS TO HELP !!!!!

Still, I do not understand why FirebaseSDK v6.20 did never cause a crash - but v6.34 does ??

To summarize:

The crash happened inside :

```Swift
       RemoteConfig.remoteConfig().fetch(withExpirationDuration: expiration) { (status, error) in
            switch status {
            case .success:
                 RemoteConfig.remoteConfig().activate { (bool, error) in
                     NotificationCenter.default.post(name: Config.Firebase.Remote.updateNotification, object: self)
                 }
            }
       }
```

**Comment by paulb777** [comment]
There have been threading fixes for RemoteConfig between 6.20.0 and 6.34.0 but it's not obvious on a quick search which one it would be.

**Comment by morganchen12** [comment]
The fix for this is https://github.com/firebase/firebase-ios-sdk/pull/6665. Previously, the threading behavior for this method was not well-defined. From Firebase 7 onward, it's guaranteed to execute on the main thread.

**Comment by paulb777** [comment]
Thanks @morganchen12 . I'll close since the fix has merged for the next release and the workaround described above should be appropriate until then.
